policies/coding-style.md

```diff
@@ -584,28 +594,93 @@ less complicated than coding pure assembly, and it doesn't provide the
 same performance guarantee across different micro-architecture. Nor is
 it portable enough to meet our multi-platform support goals.
 
-## Chapter 14:  Portability
+## Chapter 14: Portability
 
 To maximise portability the version of C defined in ISO/IEC 9899:1990
 should be used. This is more commonly referred to as C90. ISO/IEC 9899:1999
 (also known as C99) is not supported on some platforms that OpenSSL is
 used on and therefore should be avoided.
 
-## Chapter 15: Miscellaneous
+## Chapter 15: Expressions
 
-Do not use `!` to check if a pointer is NULL, or to see if a str...cmp
-function found a match.  For example, these are wrong:
+Avoid needless parentheses as far as reasonable.
+For example, do not write
+```c
+    if ((p == NULL) && (!f(((2 * x) + y) == (z++))))
+```
+but
+```c
+    if (p == NULL && !f(2 * x + y == z++)).
+```
 
+For clarity, always put parentheses when mixing `&&` and `||` operations,
+comparison operators like `<=` and `==`, and bitwise operators.
```

Comment:
This one surprised me 😕

Reply:
So now, `(a == 1 || b == 2)` is badly seen?  We have written that sort of expression all over the place until now...

Reply:
This recent addition was motivated by https://github.com/openssl/technical-policies/pull/39#discussion_r846933175.

Reply:
That comment was intended as the kind of mess C operators can create.  I don't think it's a great example to make a rule from, but I can live with it.  I have seen code along the lines of `a < b != c < d`.

I think we need some flexibility in interpretation.  Minimal parenthesis is an ideal but needs to be tempered but comprehension.

Reply:
> I think we need some flexibility in interpretation

I agree on principle, but have a hard time seeing the flexibility in "always"

Reply:
I can only agree with this sentence if the **and** is taken literally:  parentheses are only required if you are mixing operators from all three categories

- (A) comparisons [6,7] 
- (B) binary bitwise boolean operators (8,9,10)
- (C) binary logical boolean operators  (11,12)

A simple expression like `(a == 1 || b == 2 && c == 3)` which only combines comparisons (A) using logical booleans (C) should not require parentheses. (And just in case: whoever does not know that AND binds stronger than OR in C should be ashamed an reread his copy of K&R.)

![grafik](https://user-images.githubusercontent.com/11467854/166438039-5c91dda8-2671-4616-9ffa-52c5216ca025.png)

(see https://en.cppreference.com/w/c/language/operator_precedence)

Reply:
I've understood it (and I think it was meant to be) as mixing operators in the same category only. I.E. this does not apply to `a == b && c > d`. But applies to `a && b || c ` which should be written as `(a && b) || c` even if the former is equivalent according to C rules.

Reply:
> (And just in case: whoever does not know that AND binds stronger than OR in C should be ashamed an reread his copy of K&R.)

It is not a problem of shame. It is a problem of easy readability from top of your head - I.E. so you do not have to think about it.

Reply:
> So now, `(a == 1 || b == 2)` is badly seen? We have written that sort of expression all over the place until now...

Nope. This should be OK with the policy as is, it is obvious from the other examples in the coding style policy. Yeah, the sentence might be clarified under the minor policy edits policy, once we have it approved. :grin:

Reply:
Personally, I really like K&R's decision to give the AND operator precedence over the OR operator and dislike putting unnecessary parentheses there, but I wouldn't be dogmatic about it. My K&R potshot was not meant seriously.

Reply:
Personally, I really like K&R's decision to give the AND operator precedence over the OR operator and dislike putting unnecessary parentheses there, but I wouldn't be dogmatic about it. My K&R potshot was not meant seriously.

Reply:
I would have hoped that the examples make clear how to understand the sentence in question.
Adding the sentence suggested by @t8m would certainly help clarifying the intended meaning.
Should I do although this might invalidate the approvals and the votes already cast?

Reply:
Please keep it as is. I think it can be added after the minor policy edits is approved.

Reply:
> I think it can be added after the minor policy edits is approved.

It would probably change the meaning...

Reply:
The way I read it the sentence above is completely contradicting the immediately previous example.

Reply:
> > I think it can be added after the minor policy edits is approved.
> 
> It would probably change the meaning...

No, the sentence @t8m proposed to add later:
> I.E., do not depend on `&&` having higher precedence than `||`, `<=` having higher precedence than `==`, and `&` having higher precedence than `|` or `^`.

would just clarify the sentence before:

> For clarity, always put parentheses when mixing `&&` and `||` operations, comparison operators like `<=` and `==,` and bitwise operators.

Reply:
> I've understood it (and I think it was meant to be) as mixing operators in the same category only. I.E. this does not apply to `a == b && c > d`. But applies to `a && b || c ` which should be written as `(a && b) || c` even if the former is equivalent according to C rules.

Upon re-reading it, I can understand the sentence like that. But it's not perfectly clear. I suspect that we might have disputes later on depending on interpretation.

Ah well